src/Compilers/CSharp/Portable/Binder/SwitchBinder_BindPatternSwitch.cs

```diff
@@ -14,156 +14,68 @@ namespace Microsoft.CodeAnalysis.CSharp
-
-            // Bind switch expression and set the switch governing type.
-            var localDiagnostics = DiagnosticBag.GetInstance();
-            var boundSwitchExpression = BindSwitchExpressionAndGoverningType(node.Expression, originalBinder, localDiagnostics);
-            diagnostics.AddRangeAndFree(localDiagnostics);
-
-            BoundPatternSwitchLabel defaultLabel;
-            ImmutableArray<BoundPatternSwitchSection> switchSections = BindPatternSwitchSections(boundSwitchExpression, node.Sections, originalBinder, out defaultLabel, diagnostics);
-            var locals = GetDeclaredLocalsForScope(node);
-            var functions = GetDeclaredLocalFunctionsForScope(node);
-            return new BoundPatternSwitchStatement(
-                node, boundSwitchExpression,
-                locals, functions, switchSections, defaultLabel, this.BreakLabel, this);
-        }
-
-        private SourceLabelSymbol GetDefaultLabel()
-        {
-            // SwitchLabelsMap: Dictionary for the switch case/default labels.
-            // Default label(s) are indexed on a special DefaultKey object.
-
-            return FindMatchingSwitchLabel(s_defaultKey);
-        }
-
-        private static readonly object s_nullKey = new object();
-        private static object KeyForConstant(ConstantValue constantValue)
-        {
-            Debug.Assert(constantValue != (object)null);
-            return constantValue.IsNull ? s_nullKey : constantValue.Value;
-        }
-
-        private SourceLabelSymbol FindMatchingSwitchCaseLabel(ConstantValue constantValue, CSharpSyntaxNode labelSyntax)
-        {
-            // SwitchLabelsMap: Dictionary for the switch case/default labels.
-            // Case labels with a non-null constant value are indexed on their ConstantValue.
-            // Invalid case labels (with null constant value) are indexed on the label syntax.
-
-            object key;
-            if (constantValue != (object)null)
-            {
-                key = KeyForConstant(constantValue);
-            }
-            else
-            {
-                key = labelSyntax;
-            }
-
-            return FindMatchingSwitchLabel(key);
-        }
-
-        private SourceLabelSymbol FindMatchingSwitchLabel(object key)
-        {
-            Debug.Assert(key != null);
-
-            var labelsMap = LabelsByValue;
-            if (labelsMap != null)
+            get
             {
-                SourceLabelSymbol label;
-                if (labelsMap.TryGetValue(key, out label))
+                if (_isPatternSwitch == null)
                 {
-                    Debug.Assert(label != (object)null);
-                    return label;
+                    var parseOptions = _switchSyntax?.SyntaxTree?.Options as CSharpParseOptions;
+                    _isPatternSwitch =
+                        (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false &&
```

Comment:
Indeed, that is exactly what `parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false` does.